paywall/src/unlock.js/setupIframeMailbox.ts

```diff
@@ -0,0 +1,155 @@
+ *   }
+ * }
+ *
+ * @returns {Function} this function accepts a map of postMessage
+ * types to handler templates (MessageHandlerTemplate)
+ * defined above. This template will receive a postMessage sender
+ * that can be used to send a message to any iframe. It also
+ * receives all of the iframes so that they can be modified,
+ * for example by showing or hiding the iframe.
+ */
+export default function setupIframeMailbox(
+  window: PostOfficeWindow,
+  checkoutIframe: IframeType,
+  dataIframe: IframeType,
+  accountIframe: IframeType
+) {
+  // first, create the post offices for each iframe
+  const {
+    postMessage: dataPostOffice,
+    addHandler: addDataMessageHandler,
+  } = mainWindowPostOffice(
+    window,
+    dataIframe,
+    process.env.PAYWALL_URL,
+    'main window',
+    'Data iframe'
+  )
+  const {
+    postMessage: checkoutUIPostOffice,
+    addHandler: addCheckoutMessageHandler,
+  } = mainWindowPostOffice(
+    window,
+    checkoutIframe,
+    process.env.PAYWALL_URL,
+    'main window',
+    'Checkout UI'
+  )
+
+  const {
+    postMessage: accountUIPostOffice,
+    addHandler: addAccountMessageHandler,
+  } = mainWindowPostOffice(
+    window,
+    accountIframe,
+    process.env.UNLOCK_APP_URL,
+    'main window',
+    'Account UI'
+  )
+
+  // next, define the dispatcher for adding new postMessage handlers
+  const addHandler = (
+    forIframe: IframeNames,
+    type: string,
+    handler: PostMessageListener
+  ) => {
+    switch (forIframe) {
+      case 'data':
+        return addDataMessageHandler(type, handler)
+      case 'account':
+        return addAccountMessageHandler(type, handler)
+      case 'checkout':
+        return addCheckoutMessageHandler(type, handler)
+    }
+  }
+
+  // next, define the dispatcher for sending postMessage messages
+  const send = (
+    to: IframeNames,
+    type: MessageTypes,
+    payload: ExtractPayload<MessageTypes>
+  ) => {
+    switch (to) {
+      case 'data':
+        return dataPostOffice(type, payload)
+      case 'account':
+        return accountUIPostOffice(type, payload)
+      case 'checkout':
+        return checkoutUIPostOffice(type, payload)
+    }
+  }
+
+  // next, define the template handler that will be used to create the postMessage listeners
+  const createHandler = (
+    forIframe: IframeNames,
+    type: string,
+    handlerTemplate: MessageHandlerTemplate<MessageTypes>
+  ) => {
+    const listener = handlerTemplate(
+      send,
+      dataIframe,
+      checkoutIframe,
+      accountIframe
+    )
+    addHandler(forIframe, type, listener)
+  }
+
+  // finally, return the mapper that will convert a map of postMessage types to handlers
+  return (
+    forIframe: IframeNames,
+    handlers: MessageHandlerTemplates<MessageTypes>
+  ) => {
+    const keys = Object.keys(handlers) as Array<keyof typeof handlers>
+    keys.forEach(type => {
+      const handler = handlers[type]
+      handler && createHandler(forIframe, type, handler)
```

Comment:
Is the `&&` check necessary? It looks like every `handlers[type]` will have a truthy value

Reply:
because the key for the handlers type is optional (otherwise every handler would have to define a listener for every postMessage type), the type checker believes there is a possibility that handler could be null.

I spent nearly an hour trying to find a way other than this to solve it... and failed.

Reply:
Makes sense, annoying when you come across a typing corner case.